presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/WindowFilterPushDown.java

```diff
@@ -60,220 +64,149 @@ public PlanNode optimize(PlanNode plan, Session session, Map<Symbol, Type> types
+            PlanNode rewrittenSource = context.rewrite(node.getSource());
+            if (rewrittenSource instanceof RowNumberNode) {
+                RowNumberNode rowNumberNode = (RowNumberNode) rewrittenSource;
+                if (rowNumberNode.getPartitionBy().isEmpty()) {
+                    return (rowNumberNode).withLimit(Ints.checkedCast(node.getCount()));
+                }
             }
 
-            return context.defaultRewrite(node);
+            if (rewrittenSource instanceof WindowNode) {
+                WindowNode windowNode = (WindowNode) rewrittenSource;
+                if (canOptimizeWindowFunction(windowNode) && windowNode.getPartitionBy().isEmpty()) {
+                    // verify that unordered row_number window functions are replaced by RowNumberNode
+                    verify(!windowNode.getOrderBy().isEmpty());
+                    return convertToTopNRowNumber(windowNode, Ints.checkedCast(node.getCount()));
+                }
+            }
+            return context.replaceChildren(node, ImmutableList.of(rewrittenSource));
         }
 
         @Override
-        public PlanNode visitFilter(FilterNode node, RewriteContext<Constraint> context)
+        public PlanNode visitFilter(FilterNode node, RewriteContext<Void> context)
         {
-            PlanNode rewrittenSource = context.rewrite(node.getSource(), new Constraint(Optional.empty(), Optional.of(node.getPredicate())));
-            if (rewrittenSource != node.getSource()) {
-                if (rewrittenSource instanceof TopNRowNumberNode) {
-                    return rewrittenSource;
-                }
-                return new FilterNode(idAllocator.getNextId(), rewrittenSource, node.getPredicate());
-            }
-            return context.defaultRewrite(node);
-        }
-    }
+            PlanNode rewrittenSource = context.rewrite(node.getSource());
 
-    private static boolean isRowNumberSignature(Signature signature)
-    {
-        return signature.equals(ROW_NUMBER_SIGNATURE);
-    }
+            if (rewrittenSource instanceof RowNumberNode || rewrittenSource instanceof WindowNode) {
 
-    private static class Constraint
-    {
-        private final Optional<Integer> limit;
-        private final Optional<Expression> filterExpression;
+                TupleDomain<Symbol> tupleDomain = fromPredicate(metadata, session, node.getPredicate(), symbolAllocator.getTypes()).getTupleDomain();
 
-        private Constraint(Optional<Integer> limit, Optional<Expression> filterExpression)
-        {
-            this.limit = limit;
-            this.filterExpression = filterExpression;
+                if (rewrittenSource instanceof RowNumberNode) {
+                    Optional<Integer> highMarker = extractHighMarker(tupleDomain, ((RowNumberNode) rewrittenSource).getRowNumberSymbol());
+                    if (highMarker.isPresent()) {
+                        return ((RowNumberNode) rewrittenSource).withLimit(highMarker.get());
```

Comment:
You need to re-include the filter node here if it contains anything else other than the filter on the row number. Also add a test for this.